tools/autograd/gen_variable_type.py

```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
             body.append(f'check_no_requires_grad({arg_name}, "{arg_name}", "{name}");')
         return body
 
+    def emit_orignal_self_definition() -> List[str]:
+        body: List[str] = []
+        if inplace:
+            body.append('c10::optional<at::Tensor> original_self;')
+
+            all_forward_grad_cond = []
+            for derivative in fw_derivatives:
+                if derivative.required_original_self_value:
+                    all_forward_grad_cond.append(get_any_has_forward_grad_name(derivative.var_name))
+
+            if all_forward_grad_cond:
+                body.append(f'if ({" || ".join(all_forward_grad_cond)}) {{')
+                body.append('  original_self = self.clone();')
+                body.append('}')
+
+        return body
+
```

Comment:
We probably want a big note somewhere about what the structure of a (for example) VariableType::mul_ code looks like. Some things I had to remind myself about include: we need to clone `self` before the redispatch because otherwise self can get mutated